Add Webhook storage (fixes #2835)

jameslamb opened the pull request:
- [x] adds new tests (if appropriate)
- [x] add a changelog entry in the `changes/` directory (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

## What does this PR change?

This PR proposes a new type of storage, `WebHook`. `WebHook` storage is a type of storage where `build()` makes an HTTP request to store a flow and `get_flow()` makes another HTTP request to retrieve it.

This proposal is documented in #2835 and in the Prefect community Slack thread it links to.

## Why is this PR important?

#2835 has more detail on why this is valuable, but the basic thread that led to this is as follows:

1. Because `Storage` objects have to be deserializable from the JSON sent to Prefect Cloud, users cannot write their own arbitrary Python code: all storage has to use a class in `prefect`
2. Many integrations could be generalized as "issue an HTTP request to store the flow, issue another to retrieve it". This includes external services (like DropBox and Box) and custom services written by users or their companies
3. Creating a storage object that makes those requests would allow users to more easily plug into services that expose "move files over HTTP / HTTTPS", without needing a new Prefect storage class for each of those services

## Notes for reviewers

* In addition to the unit tests here, I added some integration tests to be sure this is working as expected with real services. See https://github.com/jameslamb/webhook-storage-integration-tests for examples of two integrations I wanted to test:
    - storing flows in DropBox
    - storing flows in a service where you have to first store the flow, capture an ID from that response, then later use that ID in `get_flow()`
* this implementation will only work in cases where you have long-lived credentials (like an API key) or where it's feasible to get short-lived credentials in an ad hoc way and then do a single flow run manually. It won't support, for example, any authentication that requires exchanging tokens at runtime. I think that's ok for a first version of this.
* this implementation assumes that a single request is enough to move a flow over the wire and doesn't support [multipart upload](https://stackoverflow.com/questions/33889410/proper-way-to-implement-restful-large-file-upload/50659874#50659874). I don't think that is too limiting, but I don't have a good idea of distribution of flow sizes...most of the ones I've written have been a handful of tasks and a few kB when serialized
* this implementation assumes that headers are the only place where sensitive information is stored. Is that a good assumption? While searching around, I didn't find any mainstream services that do things like  `POST service.whatever/upload?apiKey={apikey}`, but I have worked with such services in the past.
    - if reviewers think that's too limiting, I could take the approach of allowing templating in  `build_kwargs` and `get_flow_kwargs
* the `o` and `t` keys on my keyboard have some stuff under them so please keep an eye out for extra letters in my docs 😬 


Thanks for your time and consideration!

joshmeek commented:
@jameslamb just letting you know this has not been forgotten. I will take a deeper look at this sometime tomorrow!

jameslamb commented:
> @jameslamb just letting you know this has not been forgotten. I will take a deeper look at this sometime tomorrow!

thanks! no problem

jameslamb commented:
> ...I actually think this storage type could be expanded to also support the file-based storage options that the other storage types now also support as of the `0.12.5` release

ooooo I haven't looked at file-based storage yet! I found #2840 and #2944 and took a look.

Is this what you're thinking for `WebHook`?

* if `storage_as_file is True`:
    - `build_*` arguments describe an HTTP request that sends *the contents of a .py file* over the wire to some place that stores it. 
    - `get_flow_*` arguments describe an HTTP request the retrieves *the contents of a .py file*. `get_flow()` executes that request, passes that file content to `extract_flow_from_file()`

I'm happy to add that if all Storage objects are expected to have it.

joshmeek commented:
@jameslamb Yep that is exactly it! All storage types now support it so it should only need to be added to WebHook 🙂

jameslamb commented:
cool ok, I can do that!

I just looked and there is actually a `text/x-python` MIME type for that! 

* https://mediatemple.net/community/products/dv/204403964/mime-types
* https://svn.python.org/projects/python/trunk/Lib/mimetypes.py

jameslamb commented:
I've added support for `stored_as_script` in https://github.com/PrefectHQ/prefect/pull/3000/commits/59bcc06dbf5e8db1fae1154d9b838ef20804f8b8.

I'm proposing this design for file-based storage with `Webhook`:

* provide a `flow_file_path` in the constructor, a path to a `.py` file with the flow
     - that file path is not serialized, so it's never sent to Prefect Cloud. File names can leak things like usernames and other sensitive information.
* `build()` reads in the contents of the file, converts that string to bytes, and attaches it to the request as `data`
* `get_flow()` expects to receive binary data that it `.decode()`-es into a a string, which is then passed to `extract_flow_from_file()`

I  decided on this approach because, in my experience, most services that read and write files expect binary content, not `application/text`, `text/x-python`, etc. If someone asks in the future for support for writing / reading script files with text MIME types, a flag could be added to control whether or not this storage should run `.encode()` / `.decode()` on the script contents. For now, I don't think that complexity is worth it.

To test this added `stored_as_script` support, I added new unit tests (in that same commit) and tested and end-to-end   flow build, register, and run by adding a `stored_as_script=True` test to the integration tests I set up to help on this PR: https://github.com/jameslamb/webhook-storage-integration-tests#script

jameslamb commented:
I just pushed one other small change in https://github.com/PrefectHQ/prefect/pull/3000/commits/8c0e32dca59f926a2dfcbd977284f0e0d1c5590f, hope it's ok...I think I prefer `Webhook` to  `WebHook` (no capital `H`).

That is the spelling used everywhere I looked, including [Wikipedia](https://en.wikipedia.org/wiki/Webhook), [GitHub help](https://docs.github.com/en/github/extending-github/about-webhooks), [Atlassian Help](https://developer.atlassian.com/server/jira/platform/webhooks/), and [IFTTT's docs](https://ifttt.com/maker_webhooks).

jcrist commented on docs/orchestration/execution/storage_options.md:
```diff
@@ -140,6 +140,59 @@ If you do not specify a `registry_url` for your Docker Storage then the image wi
+            },
+        },
+        get_flow_http_method="POST",
+        build_secret_config={
```
I also find the `get_flow` and `build` prefixes on these kwargs a bit off. I know they correspond to the requests for the `build` and `get_flow` methods, but without the word `request` in there `build_kwargs` looks like kwargs to pass to `build` to me. Feels too tied to the interface method names and not tied to what the requests are actually doing (storing and loading bytes). Perhaps?

```
store_request_kwargs=...,
store_request_method=...,
load_request_kwargs=...,
load_request_method=...,
```

I'd use `put` and `get`, except those conflict with the http methods. Not a strong opinion, just a thought.

jameslamb replied:
I thought this was sufficiently complex that I shouldn't make it part of the first pass, but if you think it's necessary I'm happy to add it!

I'm a little worried about free-form templating everything though...that's going to be a problem if you have JSON jammed in a string, like the DropBox API requires (https://github.com/jameslamb/webhook-storage-integration-tests/blob/3bc93bf2ce4b9a0539306045f2f6a82bc3325c53/test-dropbox.py#L47). That opens you up to needing to know how to escape the right `}`, which doesn't sound fun.

maybe it would be simpler to, instead of templating individual string fields, just allow people to replace the *entire value* of any build_kwarg or get_flow_kwarg with the value of an environment variable / secret?

>  what if you need to modify the secret before transmitting (e.g. adding a Bearer prefix))

I did think about this specific case. If your API token's literal value is `abc`, there's no reason you couldn't put `Bearer abc` into an environment variable / secret, right? Without needing to have any code run to add `Bearer` to the front.

jcrist replied:
It feels non-intuitive to me to require storing a full authorization header/url/etc... in a secret to make proper use of it. If we keep the `$` prefix requirement that `string.Template` uses, that (I believe) avoids the issue of accidentally templating things that just happen to contain `{}` characters, since they won't match the regex.

jameslamb replied:
ooooo ook! I like that a lot. I'll add that to this PR.

As for the names, I feel that there's value in coupling to the method names actually. `build()` and `get_flow()` are important to understand when using a Storage object, I think, and I'd rather couple to those than invent another thing people have to reason about. But I do like adding `_request` to make that clearer. How do you feel about `build_request_*`and `get_flow_request_*`?

jcrist replied:
>  But I do like adding `_request` to make that clearer.

Makes sense to me!

jameslamb replied:
Ok I just attempted the templating thing. Awesome suggestion, I like this a lot better than the `secret_config` approach.

Commit: https://github.com/PrefectHQ/prefect/pull/3000/commits/937dd84700ba5799319da2637a593d4a61cbbf64

I also updated my integration tests and ran them to be sure it's working as expected: https://github.com/jameslamb/webhook-storage-integration-tests.

## Note for reviewers

I think it could be valuable to offer more explicit control over whether environment variables or Prefect secrets are used, to avoid issues caused by conflicting names.

I think that could be done in a backwards-compatible way in the future, by adding a `render_preferences` argument that is like `{"SOME_VARIABLE": "environment"}`,  which changes the behavior for rendering `${SOME_VARIABLE}` from "env --> secret --> error-if-absent" to "env --> error-if-absent".

I thought that complexity wasn't worth it for a first pass, but I'd be happy to add something like it if reviewers think it's a good idea.

jcrist commented on src/prefect/environments/storage/webhook.py:
```diff
@@ -0,0 +1,404 @@
+        - RuntimeError if any template value cannot be found in the environment
+            or `prefect` secrets
+    """
+    input_dict = deepcopy(input_dict)
```
I don't believe deepcopy is necessary here, since you're building a new dict.

```suggestion
```

jameslamb replied:
oh sure! Fine with me if you want to drop it. I just always do that out of an abundance of caution. I expect these dicts to be so small that the performance and memory hit from doing that is negligble. I'm ok with removing it.

jameslamb replied:
removed in https://github.com/PrefectHQ/prefect/pull/3000/commits/921e387d26e624b248de962e3fda9b5a54f1db91. I decided not to commit the suggestion in the browser, since I also needed to remove `from copy import deepcopy` and didn't want to waste CI resources by making this change in two commits

jameslamb commented:
Thanks for the thorough reviews and for teaching me the `string.Template` + magic mapping combination!
